src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -75,107 +29,156 @@ internal sealed class BuildServerConnection
-                pipeNameOpt,
-                keepAlive,
-                libEnvVariable,
-                timeoutOverride: null,
-                createServerFunc: TryCreateServerCore,
-                logger: logger,
-                cancellationToken: cancellationToken);
+                workingDirectory: workingDirectory,
+                tempDirectory: tempDirectory,
+                compilerHash: BuildProtocolConstants.GetCommitHash() ?? "",
+                requestId: requestId,
+                keepAlive: keepAlive,
+                libDirectory: libDirectory);
         }
 
-        internal static async Task<BuildResponse> RunServerCompilationCoreAsync(
-            Guid requestId,
-            RequestLanguage language,
-            List<string> arguments,
-            BuildPathsAlt buildPaths,
-            string? pipeName,
-            string? keepAlive,
-            string? libDirectory,
+        /// <summary>
+        /// Shutting down the server is an inherently racy operation.  The server can be started or stopped by
+        /// external parties at any time.
+        /// 
+        /// This function will return success if at any time in the function the server is determined to no longer
+        /// be running.
+        /// </summary>
+        internal static async Task<bool> RunServerShutdownRequestAsync(
+            string pipeName,
             int? timeoutOverride,
-            CreateServerFunc createServerFunc,
+            bool waitForProcess,
             ICompilerServerLogger logger,
             CancellationToken cancellationToken)
         {
-            if (pipeName is null)
+            if (wasServerRunning(pipeName) == false)
             {
-                throw new ArgumentException(nameof(pipeName));
+                // The server holds the mutex whenever it is running, if it's not open then the 
+                // server simply isn't running.
+                return true;
             }
 
-            if (buildPaths.TempDirectory == null)
+            try
             {
-                throw new ArgumentException(nameof(buildPaths));
+                var request = BuildRequest.CreateShutdown();
+
+                // Don't create the server when sending a shutdown request. That would defeat the 
+                // purpose a bit.
+                var response = await RunServerBuildRequestAsync(
+                    request,
+                    pipeName,
+                    timeoutOverride,
+                    tryCreateServerFunc: (_, _) => false,
+                    logger,
+                    cancellationToken).ConfigureAwait(false);
+
+                if (response is ShutdownBuildResponse shutdownBuildResponse)
+                {
+                    if (waitForProcess)
+                    {
+                        try
+                        {
+                            var process = Process.GetProcessById(shutdownBuildResponse.ServerProcessId);
+                            process.WaitForExit();
```

Comment:
Should this wait be asynchronous?

Reply:
There is a `WaitForExitAsync` method but it's only available from `net5.0` and higher. I looked at the implementation and debated copying it into our repo for the older cases. The code is complex though, has a number of races to consider and the runtime team has already done one round of fixup on it. Decided that the risk / reward isn't there for copying for downlevel platforms. Added a `#if` so that we use the `async` version when available.